compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -1567,6 +1575,97 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    fn note_type_is_not_clone_inner_expr<'b>(
+        &'b self,
+        expr: &'b hir::Expr<'b>,
+    ) -> Option<&'b hir::Expr<'b>> {
+        // When `expr` is `x` in something like `let x = foo.clone(); x`, need to recurse up to
+        // get `foo` and `clone`.
+        match expr.kind {
+            hir::ExprKind::Path(hir::QPath::Resolved(
+                None,
+                hir::Path { segments: [_], res: crate::Res::Local(binding), .. },
+            )) => {
```

Comment:
All the other arms of this match block return. Therefore you can extract the binding like so:

```rust
let binding = match expr.kind {
    hir::ExprKind::Path(hir::QPath::Resolved(
        None,
        hir::Path { segments: [_], res: crate::Res::Local(binding), .. },
    )) => {
        binding
    }
    /* other arms return */
};

Reply:
Good suggestion but it’s no longer true that all the other branches return now we’re not returning an Option. I think this is clearer as is personally, but happy to change if you still think.